pkg/koordlet/util/perf/perf_linux.go

```diff
@@ -22,15 +22,377 @@ package perf
+	if err != nil {
+		return nil, err
+	}
+	return collector, nil
+}
+
+func GetContainerPerfResult(collector *PerfGroupCollector) (map[string]float64, error) {
+	var err error
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	wg.Add(len(collector.cpus))
+	for _, cpu := range collector.cpus {
+		go func(cpu int) {
+			defer wg.Done()
+			if pcInf, ok := collector.perfCollectors.Load(cpu); ok {
+				pc := pcInf.(*perfCollector)
+				if err = pc.collect(collector.valueCh); err != nil {
+					mu.Lock()
+					err = multierr.Append(err, err)
+					mu.Unlock()
+				}
+			}
+		}(cpu)
+	}
+	wg.Wait()
+	err = multierr.Append(err, collector.cleanUp())
+	<-collector.closeCh
+
+	return collector.resultMap, err
+}
+
+func GetContainerCyclesAndInstructionsGroup(collector *PerfGroupCollector) (float64, float64, error) {
+	resMap, err := GetContainerPerfResult(collector)
+	if err != nil {
+		return 0, 0, err
+	}
+	return resMap["cycles"], resMap["instructions"], nil
+}
+
+func (c *PerfGroupCollector) cleanUp() error {
+	err := c.cgroupFile.Close()
+	if err != nil {
+		return fmt.Errorf("close cgroupFile %v, err : %v", c.cgroupFile.Name(), err)
+	}
+	close(c.valueCh)
+	return nil
+}
+
+// caller must free the memory
+func createPerfConfig(event string) (*unix.PerfEventAttr, error) {
+	// https://pkg.go.dev/cmd/cgo OOM instread of check malloc error
+	perfEventAttrPtr := C.malloc(C.ulong(unsafe.Sizeof(unix.PerfEventAttr{})))
+	C.memset(perfEventAttrPtr, 0, C.ulong(unsafe.Sizeof(unix.PerfEventAttr{})))
+	if err := pfmGetOsEventEncoding(event, perfEventAttrPtr); err != nil {
+		return nil, err
+	}
+
+	return (*unix.PerfEventAttr)(perfEventAttrPtr), nil
+}
+
+// pfmPerfEncodeArgT represents structure that is used to parse perf event name
+// into perf_event_attr using libpfm4.
+type pfmPerfEncodeArgT struct {
+	attr unsafe.Pointer
+	fstr unsafe.Pointer
+	size C.size_t
+	_    C.int // idx
+	_    C.int // cpu
+	_    C.int // flags
+}
+
+// https://man7.org/linux/man-pages/man3/pfm_get_os_event_encoding.3.html
+func pfmGetOsEventEncoding(event string, perfEventAttrPtr unsafe.Pointer) error {
+	arg := pfmPerfEncodeArgT{}
+	arg.attr = perfEventAttrPtr
+	fstr := C.CString("")
+	defer C.free(unsafe.Pointer(fstr))
+	arg.size = C.ulong(unsafe.Sizeof(arg))
+	eventCStr := C.CString(event)
+	defer C.free(unsafe.Pointer(eventCStr))
+	if err := C.pfm_get_os_event_encoding(eventCStr, C.PFM_PLM0|C.PFM_PLM3, C.PFM_OS_PERF_EVENT, unsafe.Pointer(&arg)); err != C.PFM_SUCCESS {
+		return fmt.Errorf("failed to get event encoding: %d", err)
+	}
+	return nil
+}
+
+func (p *perfCollector) collect(ch chan perfValue) error {
+	if err := p.stop(); err != nil {
+		return err
+	}
+	bufPool := BufPools[len(p.fds)+1]
+	buf := bufPool.Get().(*[]byte)
+	defer bufPool.Put(buf)
+	_, err := p.leaderFd.Read(*buf)
+	if err != nil {
+		return err
+	}
+
+	header := &perfValueHeader{}
+	reader := bytes.NewReader(*buf)
+	if err := binary.Read(reader, binary.LittleEndian, header); err != nil {
```

Comment:
is it necessary to check whether it is LittleEndian or BigEndian

Reply:
Keep same with `perf-utils`, most of the system is LittleEndian.